tla-bmcmt/src/test/scala/at/forsyte/apalache/tla/bmcmt/rules2/TestSetCupStratifiedRule.scala

```diff
@@ -0,0 +1,52 @@
+package at.forsyte.apalache.tla.bmcmt.rules2
+
+import at.forsyte.apalache.tla.bmcmt.types.CellT
+import at.forsyte.apalache.tla.bmcmt.{ArenaCell, Binding, FixedElemPtr, PureArena}
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.types.tla
+import org.junit.runner.RunWith
+import org.scalatest.funsuite.AnyFunSuite
+import org.scalatestplus.junit.JUnitRunner
+
+@RunWith(classOf[JUnitRunner])
+class TestSetCupStratifiedRule extends AnyFunSuite {
+
+  sealed case class MockRewriter(cheatyMap: Map[UID, ArenaCell]) extends Rewriter {
+    def rewrite(ex: TlaEx)(startingScope: RewriterScope): (RewriterScope, ArenaCell) =
+      (startingScope, cheatyMap(ex.ID))
+  }
+
+  test("Basic") {
+
+    val lSetCell = new ArenaCell(0, CellT.fromType1(SetT1(IntT1)))
+    val lElems = Seq(1, 2).map(new ArenaCell(_, CellT.fromType1(IntT1)))
+    val rSetCell = new ArenaCell(3, CellT.fromType1(SetT1(IntT1)))
+    val rElems = Seq(1).map(new ArenaCell(_, CellT.fromType1(IntT1))) // intentional duplicate
+
+    val arena = PureArena.empty.appendCell(lSetCell).appendCellSeq(lElems: _*).appendCell(rSetCell)
+    val arenaWithHas =
+      arena.appendHas(lSetCell, lElems.map(FixedElemPtr): _*).appendHas(rSetCell, rElems.map(FixedElemPtr): _*)
+
+    val lSet = tla.name("S", SetT1(IntT1))
+    val rSet = tla.name("T", SetT1(IntT1))
+    val cup = tla.cup(lSet, rSet).build
+
+    // We don't have other rules implemented, so we have to hack it a bit
+    val cellMap: Map[UID, ArenaCell] = cup match {
+      case OperEx(_, left, right) =>
+        Map(
+            left.ID -> lSetCell,
+            right.ID -> rSetCell,
+        )
+    }
+    val rule = new SetCupStratifiedRule(MockRewriter(cellMap))
+
+    val startScope = RewriterScope(arenaWithHas, new Binding(Map.empty))
+
+    val (RewriterScope(resultArena, _), resultCell) = rule.apply(cup)(startScope)
+
+    println(resultCell)
+    println(resultArena)
```

Comment:
Shouldn't these be assertions or something?

Reply:
Yeah, I didn't have a meaningful testing framework figured out at this point yet, mostly because if you look into the "constraint generation" part, it's not currently hooked up to `Z3SolverContext`, so there's no constraints being generated at the moment. As we add more rules, the tests will likely just assert that the post-arenas have a specific shape.

Reply:
> The design looks great! We will revisit `auxData` in the future, right?

Yes, this is mostly just a 1st draft of the interface. I already have some follow-up to this PR planned, which will show different use cases for auxData, as well as an interface super-trait, which is parameter-free.
 
> Are there meaningful tests to be added for assignment and subst?

Yes, but this PR is mostly just a sanity check on the interface, since there's not much to test atm, without meaningful constraint generation, or different arena rules.